exercises/generic_sorting/descr.md

```diff
@@ -0,0 +1,296 @@
+# Generic Sorting
+
+**Sorting** is the problem of rearranging an input sequence according to a
+given total preorder.
+
+It is well-known that, in order to sort a sequence of `n` elements, a
+**comparison-based** sorting algorithm needs `Omega(n log n)` comparisons,
+that is, at least `k * n * log n` comparisons, for a certain constant factor
+`k`.
+
+Comparison-based sorting algorithms are popular because they are
+**polymorphic** in the type of elements. That is, **they do not and cannot
+inspect** the structure of elements; they have access only to a user-supplied
+comparison function.
+
+That said, not every sorting algorithm is comparison-based. For certain types
+of elements and for certain preorders, other techniques may be available.
+**[Pigeonhole sort](https://en.wikipedia.org/wiki/Pigeonhole_sort)**,
+for instance, sorts a sequence of (bounded) integers in time
+`O(n)`. Furthermore, there are constructions by which **linear-time
+sorting algorithms can be composed**: for instance, if there is a linear-time
+algorithm that sorts values of type `'k1` according to a preorder `o1`, and if
+there is a linear-time algorithm that sorts values of type `'k2` according to
+a preorder `o2`, then it is not very difficult to construct a linear-time
+algorithm that sorts **pairs** of type `'k1 * 'k2` according to the
+lexicographic ordering `o1 * o2`.
+
+These remarks suggest that there is **a family of preorders** for which
+one can construct a linear-time sorting algorithm.
+
+In this exercise, we exhibit such a family of preorders. It takes the form of
+an algebraic data type `'k order`. A value `o` of type `'k order` is a
+**description** of a total preorder on values of type `'k`.
+For every such value `o`, we are able to give a linear-time sorting algorithm.
+This claim is substantiated by defining a function `sort` whose type is as follows:
+
+```
+  val sort: 'k order -> ('k * 'v) list -> 'v list
+```
+
+The function `sort o` accepts a list of key-value pairs,
+sorts it according to the preorder `o`,
+which is a preorder on keys,
+and returns just a list of values.
+The keys are dropped.
```

Comment:
Maybe a word about this design choice? Why don't we simply type `sort` as follows:
```
sort: 'a order -> 'a list -> 'a list
```
?

Reply:
I have added an explanation.